Commit Branch Charm Config Update

manadart opened the pull request:
## Description of change

This patch adds the crucial step of updating the charm configuration documents in the _settings_ collection when a branch is committed.

The settings documents are all written in a single transaction so that we never get into a partially applied state. 

This required adding a method to `State` in the settings module. I am not terribly fond of how this is done, but the generations requirements are exposing weaknesses in how settings is implemented in the state package.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- Bootstrap.
- `juju deploy redis`.
- `juju branch test-branch`.
- `juju config redis password=pass --branch test-branch`.
- (Optional) `config redis password --branch master` and verify there is no value.
- `juju commit test-branch`.
- `juju config redis password` shows "pass".

## Documentation changes

None.

## Bug reference

N/A

manadart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Overwrite Settings Documents Upon Committing a Generation](https://trello.com/c/BjOA1XM6/960-overwrite-settings-documents-upon-committing-a-generation)

achilleasa commented on state/settings.go:
```diff
@@ -266,6 +266,29 @@ func (s *Settings) applyChanges(changes settings.ItemChanges) {
+// TODO (manadart 2019-04-30): This is inelegant.
+// The settings abstractions are poor, but model generations really makes
+// this logic creak.
+func (st *State) WriteBulkSettings(s []Settings) error {
```
Given that the API surface for Settings is not that large couldn't you extract an interface out of it and define a type (e.g SettingsGroup) whose Write() implementation does the ops aggregation?

manadart replied:
That is how I approached it initially (minus the interface). It's in the first commit (https://github.com/juju/juju/pull/10120/commits/7ed76f88e42fab74cb774b1c5ce948171a1a1730).

But this type is not passed around, at least the with the current way I have done it. The only way an indirection would be of use is if we pushed the generation of the group into _state/modelgeneration.go_. I'd rather do that kind of logic at the API server level rather than in the `state` package, which already has too much business logic and interdependencies.

The blocker I ran into is that I couldn't test it (creating a mock `state.Database` and using it in the `Settings` instances) without exporting more of settings from the state package.

If you have an insight here, I'd be happy to hear it.

achilleasa replied:
Ah, I didn't see that because I was reviewing in the Files changed view.

Disclaimer: I don't know enough about settings so the following blurb may not apply...

Would it be possible to extract Settings out of the state package? Getting a DB handle from outside the state package is definitely a problem but if they could live in a standalone package perhaps you could have a `State.WriteSettings(Settings)` (where Settings is an interface)

Apart from this bit the code LGTM. Perhaps someone with more experience around this part of the codebase could suggest a more efficient approach to implementing this.

howbazaar replied:
I think what we really need here is something like `settingsUpdateOps` used as part of the build txn for `branch.Commit`. This way the read of the settings is part of the build txn and in sync, and the committing of the branch and the updating of the settings are all done in one database transaction.

manadart replied:
All done in the one transaction now, including assertions documents have not changed between reading and writing.

jameinel commented on state/settings.go:
```diff
@@ -266,6 +266,29 @@ func (s *Settings) applyChanges(changes settings.ItemChanges) {
+		allOps = append(allOps, ops...)
+	}
+
+	return errors.Annotate(st.db().RunTransaction(allOps), "writing multiple settings")
```
This should be turned into a 

st.db.Run(func(attempt int) {
  // do the cfg.settingsUpdateOps here
})

That allows for cases where the changes have already been applied by a different controller, etc.

jameinel commented on apiserver/facades/client/modelgeneration/modelgeneration.go:
```diff
@@ -148,19 +150,54 @@ func (api *API) CommitBranch(arg params.BranchArg) (params.IntResult, error) {
+		if err != nil {
+			return errors.Trace(err)
+		}
+		cfg, err := app.CharmSettingsWithDelta(delta)
```
My concern with this is that reading the CharmSettings should be part of the transaction, with some sort of assert that they haven't changed before we write them out again.
Otherwise concurrent updates will step on eachothers toes in ways we may not expect. (Namely, a branch that changes foo, but negate a change to 'bar' because we read the config before bar was update, and wrote it out.)
I believe the Settings code actually makes sure to only "$set: {doc.SETTING: newval}" so it doesn't touch the other values.
But that would need to be talking in terms of "deltas", not in terms of "allSettings".

Am I missing something?

howbazaar commented on apiserver/facades/client/modelgeneration/interface.go:
```diff
@@ -17,6 +18,7 @@ type State interface {
 	ControllerTag() names.ControllerTag
 	Model() (Model, error)
 	Application(string) (Application, error)
+	WriteBulkSettings([]state.Settings) error
```
I'm not sure why we need this as part of the State interface in the apiserver.

manadart replied:
Gone.

howbazaar commented on apiserver/facades/client/modelgeneration/modelgeneration.go:
```diff
@@ -148,19 +150,54 @@ func (api *API) CommitBranch(arg params.BranchArg) (params.IntResult, error) {
+// appropriate application charm settings documents and writes them back
+// to the DB. This writes the ultimate data as a single transaction,
+// so a branch is either committed without error, or there are no changes.
+func (api *API) applyBranchConfig(branch Generation) error {
```
This logic is in the wrong place.

The application of the branch config to the various applications needs to be part of the `branch.Commit` work. Otherwise we are splitting the transaction.

manadart replied:
Yep, done.

howbazaar commented on apiserver/facades/client/modelgeneration/modelgeneration_test.go:
```diff
@@ -115,18 +116,51 @@ func (s *modelGenerationSuite) TestTrackBranchSuccess(c *gc.C) {
 
-func (s *modelGenerationSuite) TestCommitGeneration(c *gc.C) {
-	defer s.setupModelGenerationAPI(c, func(ctrl *gomock.Controller, _ *mocks.MockState, mod *mocks.MockModel) {
+func (s *modelGenerationSuite) TestCommitBranchSuccess(c *gc.C) {
```
From the API server's point of view, all the branch component application needs to be in the database tier IMO.

manadart replied:
Yes; it is there now.

howbazaar commented on apiserver/facades/client/modelgeneration/interface.go:
```diff
@@ -42,6 +44,7 @@ type Generation interface {
 // Application describes application state used by the model generation API.
 type Application interface {
 	UnitNames() ([]string, error)
+	CharmSettingsWithDelta(settings.ItemChanges) (*state.Settings, error)
```
I'm not sure you are going to need this either.

manadart replied:
Gone.

manadart commented:
I've made the necessary changes. Because I removed logic I added rather than reverting commits, some of the files now have small tidy-ups unrelated to the ultimate PR theme.

I will squash all of this for merging once it is approved.

manadart commented:
This could safely land on 2.6 (currently targeting develop) if @jameinel or @howbazaar want to hit the button before I start on Monday.

manadart commented on state/modelgeneration.go:
```diff
@@ -424,6 +422,52 @@ func (g *Generation) Commit(userName string) (int, error) {
 }
 
+// assignedWithAllUnits generates a new value for the branch's
+// AssignedUnits field, to indicate the all units of changed applications
```
Gah, "that".

howbazaar replied:
I don't really understand why.

manadart replied:
This is for the history tracking requirement. The generation will then indicate all the units that were extant at the time of commit.

howbazaar commented on state/modelgeneration.go:
```diff
@@ -424,6 +422,52 @@ func (g *Generation) Commit(userName string) (int, error) {
+
+		// Assert that the settings document has not changed underneath us
+		// in addition to appending the field changes.
+		ops = append(ops, cfg.assertUnchangedOp())
```
Isn't it possible that the underlying settings have been updated to match the operations for a particular application?
In those situations, isn't it possible that there is a no-op for updating the settings for that application?

manadart replied:
Yes, true. I will omit the check when there are no settings modification ops generated.

howbazaar commented on state/modelgeneration.go:
```diff
@@ -424,6 +422,52 @@ func (g *Generation) Commit(userName string) (int, error) {
+// single transaction.
+func (g *Generation) commitConfigTxnOps() ([]txn.Op, error) {
+	var ops []txn.Op
+	for appName, delta := range g.Config() {
```
Where is the op that inserts the commit into the history and removes the branch?

manadart replied:
See my comment regarding generation ID above. Once completed it is a model generation and is no longer "in-flight".

manadart commented on state/modelgeneration.go:
```diff
@@ -400,21 +400,19 @@ func (g *Generation) Commit(userName string) (int, error) {
-					}},
-				},
+		ops = append(ops, txn.Op{
+			C:      generationsC,
```
The document gets a generation ID and a completion date once committed, which indicates that it is no longer an in-flight branch.

manadart commented:
!!build!!

manadart commented:
@howbazaar is this OK to land with the changes? I rather think this will undergo re-design in any case.

manadart commented:
$$merge$$
